engine/version/version_controller.hpp

```diff
@@ -70,6 +70,63 @@ class SnapshotList {
 // The snapshots include temporal snapshots that cached by each access thread of
 // kvdk instance, and a global snapshot list that actively created by user
 class VersionController {
+ public:
+  class LocalToken {
+    VersionController* owner_{nullptr};
+    TimeStampType ts_{};
+
+   public:
+    LocalToken(VersionController* o) : owner_{o} {
+      owner_->HoldLocalSnapshot();
+      ts_ = owner_->version_thread_cache_[access_thread.id]
+                .holding_snapshot.timestamp;
+    };
+    LocalToken(LocalToken const&) = delete;
+    LocalToken& operator=(LocalToken const&) = delete;
+    LocalToken(LocalToken&& other) { *this = std::move(other); }
+    LocalToken& operator=(LocalToken&& other) {
+      kvdk_assert(owner_ == nullptr, "");
+      kvdk_assert(other.owner_ != nullptr, "");
+      std::swap(owner_, other.owner_);
+      std::swap(ts_, other.ts_);
+      return *this;
+    }
+    ~LocalToken() {
+      if (owner_ != nullptr) {
+        owner_->ReleaseLocalSnapshot();
+      }
+    }
+    TimeStampType Timestamp() { return ts_; }
+  };
+
+  class GlobalToken {
```

Comment:
This can be dangerous, the snapshot may contains half-written data of batch write

Reply:
Let's discuss this in detail.